README.md

```diff
@@ -1,115 +1,174 @@
+StableHashTraits.transformer(::Type{<:MyType}) = Transformer(x -> (; x.data);
+                                                             preserves_structure=true)
 a = MyType(read("myfile.txt"), Dict{Symbol, Any}(:read => Dates.now()))
 b = MyType(read("myfile.txt"), Dict{Symbol, Any}(:read => Dates.now()))
-stable_hash(a; version=2) == stable_hash(b; version=2) # true
+stable_hash(a; version=3) == stable_hash(b; version=3) # true
+```
+<!--END_EXAMPLE-->
+
+Useres can define a method of `transformer` to customize how an object is hashed. It should return a function wrapped in `Transformer`. During hashing, this function is called and its result is the value that is actually hashed. (The `preserves_structure` keyword shown above is an optional flag that can be used to further optimize performance of your transformer in some cases; you can do this any time the function is type stable, but some type-unstable functions are also possible. See the documentation for details).
+
+StableHashTraits aims to guarantee a stable hash so long as you only upgrade to non-breaking versions (e.g. `StableHashTraits = "1"` in `[compat]` of `Project.toml`); any changes in an object's hash in this case would be considered a bug.
+
+> [!WARNING]
+> Hash versions 3 constitutes a substantial redesign of StableHashTraits so as to avoid reliance on some unstable Julia internals. Hash versions 1 and 2 are deprecated and will be removed in a soon-to-be released StableHashTraits 2.0. Hash version 3 will remain unchanged in this 2.0 release. Hash version 1 is the default version if you don't specify a version.
+
+<!--The START_ and STOP_ comments are used to extract content that is also repeated in the documentation-->
+<!--START_OVERVIEW-->
+## Use Case and Design Rationale
+
+StableHashTraits is designed to be used in cases where there is an object you wish to serialize in a content-addressed cache. How and when objects pass the same input to a hashing algorithm is meant to be predictable and well defined, so that the user can reliably define methods of `transformer` to modify this behavior.
+
+## What gets hashed?
+
+By default, an object is hashed according to its `StructType` (ala
+[SructTypes](https://github.com/JuliaData/StructTypes.jl)), and this can be customized using
+[`StableHashTraits.transformer`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.transformer).
+
+Hashing makes use of [`stable_type_name`](https://beacon-biosignals.github.io/StableHashTraits.jl/stable/api/#StableHashTraits.stable_type_name) which generates `string(parentmodule(T)) * "." * string(nameof(T))` for type `T`, with a few additional regularizations to ensure e.g. `Core.` values become `Base.` values (as what is in `Core` vs. `Base` changes across julia versions). This function also ensures that no anonymous values (those that include `#`) are hashed, as these are not stable across sessions.
```

Comment:
Proposal: very conservative by default, but easy to opt-in to different behavior by defining `StableHashTraits.transformer` methods:

* types by default error in value position
* functions by default error 
* singleton types by default error

Examples:

```julia
f(x) = x+1
# ideally: per-function
StableHashTraits.transformer(::typeof(f), ::MyContext) = Transformer(ParentModuleNameOfFieldHasher)

# if you want to, can opt-in for all functions, but can be risky (if they move submodules then hash changes):
StableHashTraits.transformer(f::T, ::MyContext) where {T <: Function} = Transformer(ParentModuleNameOfFieldHasher)

# can support types or a family of types or all types:
StableHashTraits.transformer(::Type{T}, ::MyContext) where {T <: MyAbstractType} = Transformer(ParentModuleNameOfHasher)
```

Rationale:
* hash implementers unfortunately need to think carefully about how to hash tricky things like types and functions, and when the hash should or should not change, so error by default
* packages that want to support StableHashTraits first-class can define `StableHashTraits.transformer` methods (without a context) for their functions/types, to declare how they should be hashed (and they should test their hashes don't change when they don't want them to)
* users can define `StableHashTraits.transformer` with a context to define how upstream functions/types are hashed (without piracy)